dGame/dBehaviors/BasicAttackBehavior.cpp

```diff
@@ -89,79 +112,119 @@ void BasicAttackBehavior::Handle(BehaviorContext* context, RakNet::BitStream* bi
 		this->m_OnSuccess->Handle(context, bitStream, branch);
 		break;
+	case eBasicAttackSuccessTypes::FAILARMOR:
+		this->m_OnFailArmor->Handle(context, bitStream, branch);
+		break;
 	default:
-		Game::logger->LogDebug("BasicAttackBehavior", "Unknown success state (%i)!", successState);
+		if (static_cast<eBasicAttackSuccessTypes>(successState) != eBasicAttackSuccessTypes::FAILIMMUNE) {
+			Game::logger->LogDebug("BasicAttackBehavior", "Unknown success state (%i)!", successState);
+			return;
+		}
+		this->m_OnFailImmune->Handle(context, bitStream, branch);
 		break;
 	}
-
-	bitStream->SetReadOffset(baseAddress + allocatedBits);
 }
 
 void BasicAttackBehavior::Calculate(BehaviorContext* context, RakNet::BitStream* bitStream, BehaviorBranchContext branch) {
-	auto* self = EntityManager::Instance()->GetEntity(context->originator);
-	if (self == nullptr) {
-		Game::logger->LogDebug("BasicAttackBehavior", "Invalid self entity (%llu)!", context->originator);
-		return;
-	}
-
 	bitStream->AlignWriteToByteBoundary();
 
 	const auto allocatedAddress = bitStream->GetWriteOffset();
 
-	bitStream->Write(uint16_t(0));
+	bitStream->Write<uint16_t>(0);
 
 	const auto startAddress = bitStream->GetWriteOffset();
 
-	bitStream->Write0(); // Blocked
-	bitStream->Write0(); // Immune
-	bitStream->Write1(); // Success
+	DoBehaviorCalculation(context, bitStream, branch);
+
+	const auto endAddress = bitStream->GetWriteOffset();
+	const uint16_t allocate = endAddress - startAddress + 1;
 
-	if (true) {
-		uint32_t unknown3 = 0;
-		bitStream->Write(unknown3);
+	bitStream->SetWriteOffset(allocatedAddress);
+	bitStream->Write(allocate);
+	bitStream->SetWriteOffset(startAddress + allocate);
+}
 
-		auto damage = this->m_MinDamage;
-		auto* entity = EntityManager::Instance()->GetEntity(branch.target);
+void BasicAttackBehavior::DoBehaviorCalculation(BehaviorContext* context, RakNet::BitStream* bitStream, BehaviorBranchContext branch) {
+	auto* targetEntity = EntityManager::Instance()->GetEntity(branch.target);
+	if (!targetEntity) {
+		Game::logger->Log("BasicAttackBehavior", "Target entity %llu is null!", branch.target);
+		return;
+	}
 
-		if (entity == nullptr) {
-			damage = 0;
-			bitStream->Write(damage);
-			bitStream->Write(false);
-		} else {
-			bitStream->Write(damage);
-			bitStream->Write(true);
+	auto* destroyableComponent = targetEntity->GetComponent<DestroyableComponent>();
+	if (!destroyableComponent || !destroyableComponent->GetParent()) {
+		Game::logger->Log("BasicAttackBehavior", "No destroyable component on %llu", branch.target);
+		return;
+	}
 
-			auto* destroyableComponent = entity->GetComponent<DestroyableComponent>();
-			if (damage != 0 && destroyableComponent != nullptr) {
-				PlayFx(u"onhit", entity->GetObjectID(), 1);
-				destroyableComponent->Damage(damage, context->originator, context->skillID, false);
-				context->ScheduleUpdate(branch.target);
-			}
+	const bool isBlocking = destroyableComponent->GetAttacksToBlock() > 0;
```

Comment:
Should we decrease this value when evaluating this behavior?

Reply:
Good catch.  My thought process was this is handled in the Damage method in DestroyableComponent, so I completely missed this.